extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -173,23 +190,86 @@ public int executeUpdate(String query) throws SQLException {
-    try (Connection connection = borrowConnection()) {
-      boolean autoCommit = connection.getAutoCommit();
-      connection.setAutoCommit(false);
-      boolean success = false;
+    withRetries(
+        () -> {
+          try (Connection connection = borrowConnection()) {
+            boolean autoCommit = connection.getAutoCommit();
+            boolean success = false;
+            connection.setAutoCommit(false);
+            try {
+              try (Statement statement = connection.createStatement()) {
+                success = callback.execute(statement);
+              }
+            } finally {
+              if (success) {
+                connection.commit();
+              } else {
+                connection.rollback();
+              }
+              connection.setAutoCommit(autoCommit);
+            }
+          }
+          return null;
+        });
+  }
+
+  private boolean isRetryable(SQLException e) {
+    String sqlState = e.getSQLState();
+
+    if (sqlState != null) {
+      return sqlState.equals(DEADLOCK_SQL_CODE)
+          || // Deadlock detected
+          sqlState.equals(SERIALIZATION_FAILURE_SQL_CODE); // Serialization failure
+    }
+
+    // Additionally, one might check for specific error messages or other conditions
+    return e.getMessage().contains("connection refused")
+        || e.getMessage().contains("connection reset");
+  }
+
+  public <T> T withRetries(Operation<T> operation) throws SQLException {
+    int attempts = 0;
+    // maximum number of retries.
+    int maxAttempts = relationalJdbcConfiguration.maxRetries().orElse(1);
+    // How long we should try, since the first attempt.
+    long maxDuration = relationalJdbcConfiguration.maxDurationInMs().orElse(5000L);
+    // How long to wait before first failure.
+    long delay = relationalJdbcConfiguration.initialDelayInMs().orElse(100L);
+
+    // maximum time we will retry till.
+    long maxRetryTime = Clock.systemUTC().millis() + maxDuration;
+
+    while (attempts < maxAttempts) {
       try {
-        try (Statement statement = connection.createStatement()) {
-          success = callback.execute(statement);
+        return operation.execute();
+      } catch (SQLException e) {
+        attempts++;
+        long timeLeft = Math.max((maxRetryTime - Clock.systemUTC().millis()), 0L);
+        if (attempts >= maxAttempts || !isRetryable(e) || timeLeft == 0) {
+          String exceptionMessage =
+              String.format(
+                  "Failed due to %s, after , %s attempts and %s milliseconds",
+                  e.getMessage(), attempts, maxDuration);
+          throw new SQLException(exceptionMessage, e.getSQLState(), e.getErrorCode());
         }
-      } finally {
-        if (success) {
-          connection.commit();
-        } else {
-          connection.rollback();
+        // Add jitter
+        long timeToSleep = Math.min(timeLeft, delay + (long) (random.nextFloat() * 0.2 * delay));
```

Comment:
suggestion: use two bounds: minDelay (start at 10ms) and maxDelay (start 100ms). Jitter between them. Grow both exponentially.

I believe this approach would be easier for users to understand and predict behaviour and will also offer more tuning flexibility.

Reply:
Noting: there's `org.apache.polaris.persistence.commits.retry` package in #1189 for a _tested_ retry-loop. There's a _LOT_ of nuance here.

Reply:
I think we going more toward implementing our own retry lib, but for JDBC specially type of one shot queries (non transactional) what we have should not require this at these knobs at moment IMHO.
Definitely if the retry loop package is available, when NO-SQL gets in, I am happy to refactor it to use that instead of this basic impl.

Reply:
I'm ok with leaving this retrier refactoring / unification til later.